numpy/lib/nanfunctions.py

```diff
@@ -644,7 +678,12 @@ def nanprod(a, axis=None, dtype=None, out=None, keepdims=np._NoValue):
 
     """
     a, mask = _replace_nan(a, 1)
-    return np.prod(a, axis=axis, dtype=dtype, out=out, keepdims=keepdims)
+
+    if np.isnan(initializer):
```

Comment:
`isnan` doesn't work for `np._NoValue`. I'm not sure it's clear that nanprod` should treat initializer specially, especially since you don't document this. Keep it simple for now, without this check, I'd say.